\documentclass[plainarticle,zihtitle,english,final,hyperref,utf8]{zihpub}
\usepackage{setspace}
\author{Ankur Sharma}
\title{Performance Tuning \& Parallelisation of Inchworm}
\bibfiles{doku}
\birthday{25. March 1991}
\placeofbirth{Allahabad, India}
\betreuer{Prof. Dr. Wolfgang E. Nagel}
\begin{document}
\listoffigures
\newpage
\section{Introduction} 
The high demand of digging deeper through the structural characteristic and behaviour of Deoxyribonucleic Acid (DNA) and Ribonucleic Acid (RNA) has driven the development of highly efficient and high-throughput sequencing, which is better known as Next Generation Sequencing (NGS). The sequencing of cDNA or the RNA sequencing is a technology that exploits the properties of next-generation sequencing to demonstrate the snapshot of RNA presence and quality from a genome at given moment of time. A variety of parallel assembly programs have been developed as an open source project as well as commercial distribution in the past few decades. Although most of these programs have been successful in assembling genomes, transcriptome assembly depicted some unique challenges in almost every phase of the program starting from the performance to the efficiency of results generated by them. \emph{Trinity} is one of those assemblers that is highly efficient and presents a novel method for efficient and robust de novo reconstruction of transcriptomes from RNA-seq data. Careful analysis of \emph{Trinity} revealed a lot of performance bottlenecks in the \emph{Inchworm} phase that could be resolved in order to generate better results in a more efficient way. This document thus discusses the performance optimisations and parallel master slave approach of computing sequence assemblies deployed in \emph{Inchworm} that boosted the performance to a great extent compared to the original stable implementation.
\subsection{Biological Background}
The detailed analysis of  an assembly tool like \emph{Inchworm} involves familiarity with a lot of biological terms and techniques. This section discusses some of those key biological elements that play crucial role in understanding the implementation and a deeper insight of how the changes in the assembly algorithm is effecting the results. In further sections, some of the important terms frequently used in this documentation are discussed:
\subsubsection{Key Terms}
\paragraph{Deoxyribonucleic Acid (DNA)}: It is a molecule that encodes the genetic information and instructions used in the development and functioning of all living organisms and even many viruses. Hence it is a well defined and suited biological information storage system. It consists of a double stranded structure coiled around each other in an anti parallel fashion. Both these strands store the same information which is replicated when the two strands get separated.
\begin{figure}[h]
\center
\includegraphics[scale=0.07]{DNA}
\hspace{50pt}
\includegraphics[scale=0.6]{RNA}
\caption{Double stranded DNA and single stranded RNA molecule}
\end{figure}\\
\paragraph{Ribonucleic Acid (RNA)}: It is a large biological molecule that performs some essential roles in the coding, decoding, regulation and expression of genes. Unlike the DNA molecule, RNA has a single stranded structure and has a much shorter chain of nucleotides. Some RNA molecules also play an active role within cells by catalysing most of the biological reactions, controlling gene expressions, or sensing and communicating responses to cellular signals. 
\paragraph{Transcriptomes}: The transcriptomes refer to a set consisting of all the RNA molecules including mRNA, rRNA, tRNA and other non coding RNA produced in one or group of cells.
\paragraph{Reads}: These are a short sequences that represents the DNA sequencing data and is generated by very expensive high-throughput sequencers and are typically less than 1000 base pairs in length.
\paragraph{k-mer}: It basically refers to n-tuple of nucleic acid and amino acid sequences that can be used to identify certain regions within biomolecules like RNA. Either they as such or their statistics giving discrete probability distribution of a number of possible k-mer combinations can be used for finding regions of interest within a molecule. These sequences (string) consist of a tuple of length k with 4 different type of elements which are A (Adenine), T (Thymine), C (Cytosine) and G (Guanine).
\paragraph{Contig}: A contig is a set of overlapping DNA segments that together represent a point of interest in a DNA molecule. In bottom-up sequencing projects, a contig refers to overlapping sequence data (reads) and in a top-down sequencing projects, they refers to the overlapping clones that form a physical map of the genome that is used to guide sequencing and assembly. Contigs can thus refer both to overlapping DNA sequence and to overlapping fragments contained in clones depending on the context.
\paragraph{Sequence Assembly}: Sequence assembly refers to aligning and merging fragments of a much longer DNA sequence in order to reconstruct the original sequence. This is needed as DNA sequencing technology cannot read whole genomes in one go, but rather reads small pieces of between 20 and 30000 bases, depending on the technology used. Typically the short fragments, called reads, result from shotgun sequencing genomic DNA, or gene transcript. 
\subsection{Trinity}
\emph{Trinity} assembler, developed at the Broad Institute and the Hebrew University of Jerusalem, represents a method for \emph{de novo} assembly of full-length transcripts that uses construction and analysis of sets of \emph{de Bruijn} graph. It fully reconstructs large fraction of transcripts, including spliced isoforms and transcripts from recently duplicated genes. \emph{Trinity} partitions the sequence data into many individual \emph{de Bruijn} graphs, each representing the transcriptional complexity at at a given gene or locus, and then processes each graph independently to extract full-length splicing isoforms and to tease apart transcripts derived from paralogous genes.  
\begin{figure}[h]
\center
\includegraphics[scale=0.55]{Trinity}
\caption{Structure of trinity}
\end{figure}
\subsubsection{Structure}
\emph{Trinity} has an excellent modular structure which distributes the entire assembly into three different pipelined phases. This specific design of \emph{Trinity} allows us to analyse it in an efficient manner by decomposing the intermediate results of different phases and running only specific module to be tested.
\begin{figure}[h]
\center
\includegraphics[scale=0.5]{Trinity_phases}
\caption{\emph{Trinity} - Workflow}
\end{figure} 
\subsubsection{Phases}
\emph{Trinity} consists of three modules namely \emph{Inchworm}, \emph{Chrysalis} and \emph{Butterfly}
\subsection{Applications}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Inchworm}
\subsection{Phases}
\subsubsection{Parsing}
\subsubsection{Sorting}
\subsubsection{Pruning}
\subsubsection{Assembling}
\subsection{Implementation}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Performance Analysis}
\subsection{Datasets}
\subsection{Hardware \& Software Environment}
\subsection{Phase Analysis of Inchworm}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Performance Optimisation}
\subsection{Parsing}
\subsubsection{Memory Mapping}
\subsubsection{Observed Improvements}
\subsection{Pruning}
\subsubsection{Parallel Pruning}
\subsubsection{Performance Gain}
\subsection{Assembly}
\subsubsection{Master-Slave Approach}
\subsubsection{Benefits}
\subsection{Overall Improvements}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Conclusion}
\subsection{Future work}

\end{document}
